### PR TITLE
Test case for image validation

### PR DESCRIPTION
**Pull request by gicmo**
Add a small test case. Uncovered already two tiny glitches:
```
FuPluginThunderbolt-DEBUG: expected image validation error [ctl, ctl]: Given location is outside of the given FW ((null))
FuPluginThunderbolt-DEBUG: expected image validation error [dev, dev]: The FW image file is for a host controller
FuPluginThunderbolt-DEBUG: expected image validation error [bad, dev]: The FW image file is for a host controller
FuPluginThunderbolt-DEBUG: expected image validation error [ctl, bad]: Given location is outside of the given FW ((null))
FuPluginThunderbolt-DEBUG: expected image validation error [bad, bad]: Given location is outside of the given FW ((null))
```
NB: the `(null)` and the "for a host controller". Added two simple fixes for those. Now they read:

```
FuPluginThunderbolt-DEBUG: expected image validation error [ctl, ctl]: Given location is outside of the given FW (N/A)
FuPluginThunderbolt-DEBUG: expected image validation error [dev, dev]: The FW image file is for a device controller
FuPluginThunderbolt-DEBUG: expected image validation error [bad, dev]: The FW image file is for a device controller
FuPluginThunderbolt-DEBUG: expected image validation error [ctl, bad]: Given location is outside of the given FW (N/A)
FuPluginThunderbolt-DEBUG: expected image validation error [bad, bad]: Given location is outside of the given FW (N/A)
```
Also, three cases we return `FWUPD_ERROR_READ`, which I am not sure is better then just `FWUPD_ERROR_INVALID_FILE`. If we pass host controller image for the firmware we also don't get the "The FW image file is for a host controller" error but the "Given location is outside of the given FW". Same for the case when we pass invalid data. Maybe we should prefix those, a la "Invalid firmware file: Given location is outside .."?

### CODE REVIEW

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 1]
```diff
@@ -518,7 +518,7 @@ fu_plugin_thunderbolt_validate_image (GBytes  *controller_fw,
 		g_set_error (error,
 			     FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
 			     "The FW image file is for a %s controller",
-			     is_host ? "device" : "host");
+			     is_host ? "host" : "device");
```
Comment (ybernat) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: `is_host` means that the controller is a host controller, so if the image isn't compatible it means that the image if for a device.

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 2]
```diff
@@ -94,7 +94,7 @@ read_location (const FuThunderboltFwLocation  *location,
 		g_set_error (error,
 			     FWUPD_ERROR, FWUPD_ERROR_READ,
 			     "Given location is outside of the given FW (%s)",
-			     location->description);
+			     location->description ? location->description : "N/A");
```
Comment (ybernat) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: Maybe it worth adding description to the other `FuThunderboltFwLocation` instead? Because `N/A` isn't much better than `null`.

File: plugins/thunderbolt/fu-thunderbolt-image.c [hunk 3]
```diff
@@ -94,7 +94,7 @@ read_location (const FuThunderboltFwLocation  *location,
 		g_set_error (error,
 			     FWUPD_ERROR, FWUPD_ERROR_READ,
```
Comment (ybernat) [inline plugins/thunderbolt/fu-thunderbolt-image.c]: I assume you are talking about this error code.
This function is generic both for the controller itself and the image, this is why I chose the more generic `FWUPD_ERROR_READ` instead of `FWUPD_ERROR_INVALID_FILE`, as the latter doesn't make any sense if the read from the controller is what failed.
We can add a flag to `FuThunderboltFwObject` to tell if this is a controller or image file and return the relevant error code depending on this flag.

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 1 above)
Comment (gicmo) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: @ybernat maybe I got just confused. Both test firmware blobs are for a host controller then?

**Comment by ybernat** [comment]
Thanks for adding the tests! The more the merrier!

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 1 above)
Comment (ybernat) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: Yes. One is the image file as-is and the other is as it's read from the controller (after removal of the headers). I admit that I should have add a description. Sorry for that.

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 2 above)
Comment (gicmo) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: You know this code so much better, I would let you decided if you can fill in more relevant information in the `.description`. I am just not a fan of passing NULL pointers to g_print (I remember this used to crash on Sun).

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 1 above)
Comment (gicmo) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: Ah, now it makes sense. I guess we cant really test the device case then. Let me drop that commit and rename some variables ;)

Comment (ybernat) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: If you want one of them to look like it's (for) a device, change the second bit in location 0x10 or 0x11 (depends on which file you are changing). Currently it's 0x02, so the second bit is set, which means "a host controller". (see [line 491](https://github.com/hughsie/fwupd/pull/214/files/c65072211c75d31f6ca1b353313009b513df038f#diff-0d02b79a376b5c204ecf747b7877cbfaL491) and [line 498](https://github.com/hughsie/fwupd/pull/214/files/c65072211c75d31f6ca1b353313009b513df038f#diff-0d02b79a376b5c204ecf747b7877cbfaL498)

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 3 above)
Comment (gicmo) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: That also makes sense. I am wondering what would be a good error code if the verification for controller itself fails. None really seem to fit properly.

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 2 above)
Comment (ybernat) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: Fair enough.

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 1 above)
Comment (gicmo) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: You guys think it is worth generating that file on-the-fly for the test?

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 3 above)
Comment (ybernat) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: >  if the verification for controller itself fails

Not sure what it means. The controller is always valid, otherwise it wouldn't work (or would be in safe-mode). If there is a read error, this looks like the relevant error.
Maybe you meant if we find there is no DROM or DRAM sections when they're expected? [line 538](https://github.com/hughsie/fwupd/pull/214/files/c65072211c75d31f6ca1b353313009b513df038f#diff-0d02b79a376b5c204ecf747b7877cbfaL538)
I used `FWUPD_ERROR_READ` because I can't imaging any other reason for this to happen. Wrong image flashed to the controller by the OEM on manufacturing line?

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 1 above)
Comment (ybernat) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: Fuzzing? :)

Maybe just prepare a few interesting tests in advance (e.g. change location that matters, change location that shouldn't matter, change location that is masked in the bits inside/outside the mask) and save them as the data set.

File: plugins/thunderbolt/fu-thunderbolt-image.c (hunk 3 above)
Comment (gicmo) [reply plugins/thunderbolt/fu-thunderbolt-image.c]: Sorry, yeah, that is what I meant. Seems indeed like a corner case, maybe not worth thinking too much about.

**Comment by gicmo** [comment]
@ybernat does that look ok now? I think I will add the different test data images in a extra PR.

File: plugins/thunderbolt/fu-self-test.c [hunk 4]
```diff
@@ -997,6 +998,96 @@ test_tree (ThunderboltTest *tt, gconstpointer user_data)
...
+	g_assert_cmpint (val, ==, VALIDATION_FAILED);
+	g_debug("expected image validation error [bad, fwi]: %s", error->message);
+	g_clear_error(&error);
+
+	/* both bad */
+	val = fu_plugin_thunderbolt_validate_image (bad_data, bad_data, &error);
+	g_assert_error (error, FWUPD_ERROR, FWUPD_ERROR_READ);
+	g_assert_cmpint (val, ==, VALIDATION_FAILED);
+	g_debug("expected image validation error [bad, bad]: %s", error->message);
+	g_clear_error(&error);
```
Comment (hughsie) [inline plugins/thunderbolt/fu-self-test.c]: I feel awful for nitpicking, but space before backet.

Comment (gicmo) [reply plugins/thunderbolt/fu-self-test.c]: I blame my editor, it should feel aweful, not you! ;)
